Tests/NIOTests/ChannelPipelineTest.swift

```diff
@@ -402,4 +446,106 @@ class ChannelPipelineTest: XCTestCase {
+        XCTAssertNoThrow(try channel.pipeline.add(handler: IndexWritingHandler(2)).wait())
+        XCTAssertNoThrow(try channel.pipeline.add(handler: IndexWritingHandler(3), after: firstHandler).wait())
+
+        channel.assertReadIndexOrder([1, 3, 2])
+        channel.assertWriteIndexOrder([2, 3, 1])
+    }
+
+    func testAddBefore() {
+        let channel = EmbeddedChannel()
+        defer {
+            XCTAssertNoThrow(try channel.finish())
+        }
+
+        let secondHandler = IndexWritingHandler(2)
+        XCTAssertNoThrow(try channel.pipeline.add(handler: IndexWritingHandler(1)).wait())
+        XCTAssertNoThrow(try channel.pipeline.add(handler: secondHandler).wait())
+        XCTAssertNoThrow(try channel.pipeline.add(handler: IndexWritingHandler(3), before: secondHandler).wait())
+
+        channel.assertReadIndexOrder([1, 3, 2])
+        channel.assertWriteIndexOrder([2, 3, 1])
+    }
+
+    func testAddAfterLast() {
+        let channel = EmbeddedChannel()
+        defer {
+            XCTAssertNoThrow(try channel.finish())
+        }
+
+        let secondHandler = IndexWritingHandler(2)
+        XCTAssertNoThrow(try channel.pipeline.add(handler: IndexWritingHandler(1)).wait())
+        XCTAssertNoThrow(try channel.pipeline.add(handler: secondHandler).wait())
+        XCTAssertNoThrow(try channel.pipeline.add(handler: IndexWritingHandler(3), after: secondHandler).wait())
+
+        channel.assertReadIndexOrder([1, 2, 3])
+        channel.assertWriteIndexOrder([3, 2, 1])
+    }
+
+    func testAddBeforeFirst() {
+        let channel = EmbeddedChannel()
+        defer {
+            XCTAssertNoThrow(try channel.finish())
+        }
+
+        let firstHandler = IndexWritingHandler(1)
+        XCTAssertNoThrow(try channel.pipeline.add(handler: firstHandler).wait())
+        XCTAssertNoThrow(try channel.pipeline.add(handler: IndexWritingHandler(2)).wait())
+        XCTAssertNoThrow(try channel.pipeline.add(handler: IndexWritingHandler(3), before: firstHandler).wait())
+
+        channel.assertReadIndexOrder([3, 1, 2])
+        channel.assertWriteIndexOrder([2, 1, 3])
+    }
+
+    func testAddAfterWhileClosed() {
+        let channel = EmbeddedChannel()
+        defer {
+            _ = try? channel.finish()
+        }
+
+        let handler = IndexWritingHandler(1)
+        XCTAssertNoThrow(try channel.pipeline.add(handler: handler).wait())
+        XCTAssertNoThrow(try channel.close().wait())
+        (channel.eventLoop as! EmbeddedEventLoop).run()
+
+        do {
+            try channel.pipeline.add(handler: IndexWritingHandler(2), after: handler).wait()
+            XCTFail("Did not throw")
+        } catch ChannelError.ioOnClosedChannel {
+            // all good
+        } catch {
+            XCTFail("Got incorrect error: \(error)")
+        }
+    }
+
+    func testAddBeforeWhileClosed() {
+        let channel = EmbeddedChannel()
+        defer {
+            _ = try? channel.finish()
```

Comment:
we should assert the return value and not swallow exceptions

Reply:
We have to swallow exceptions here. `finish` will throw if the `close` future errors, which it will, because the channel is already closed. I know this because I closed it in the test. ;)

Reply:
Well what I meant was that if we expect it to throw we should assert this